Extension/src/Automation/Async/manual-promise.ts

```diff
@@ -0,0 +1,102 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All Rights Reserved.
+ * See 'LICENSE' in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+/**
+ * A manually (or externally) controlled asynchronous Promise implementation
+ */
+export class ManualPromise<T = void> implements Promise<T> {
+    readonly [Symbol.toStringTag] = 'Promise';
+    private promise: Promise<T>;
+
+    /**
+     * Attaches callbacks for the resolution and/or rejection of the Promise.
+     * @param onfulfilled The callback to execute when the Promise is resolved.
+     * @param onrejected The callback to execute when the Promise is rejected.
+     * @returns A Promise for the completion of which ever callback is executed.
+     */
+    then<TResult1 = T, TResult2 = never>(onfulfilled?: | ((value: T) => TResult1 | PromiseLike<TResult1>) | null | undefined, onrejected?: | ((reason: any) => TResult2 | PromiseLike<TResult2>) | null | undefined): Promise<TResult1 | TResult2> {
+        return this.promise.then(onfulfilled, onrejected);
+    }
+
+    /**
+     * Attaches a callback for only the rejection of the Promise.
+     * @param onrejected The callback to execute when the Promise is rejected.
+     * @returns A Promise for the completion of the callback.
+     */
+    catch<TResult = never>(onrejected?: | ((reason: any) => TResult | PromiseLike<TResult>) | null | undefined): Promise<T | TResult> {
+        return this.promise.catch(onrejected);
+    }
+
+    /**
+     * Attaches a callback that is invoked when the Promise is settled (fulfilled or rejected). The
+     * resolved value cannot be modified from the callback.
+     * @param onfinally The callback to execute when the Promise is settled (fulfilled or rejected).
+     * @returns A Promise for the completion of the callback.
+     */
+    finally(onfinally?: (() => void) | null | undefined): Promise<T> {
+        return this.promise.finally(onfinally);
+    }
+
+    /**
+     * A method to manually resolve the Promise.
+     */
+    public resolve: (value?: T | PromiseLike<T> | undefined) => void = (v) => {
+        void v; /* */
```

Comment:
Since `resolve` and `reject` are set in the constructor, should we change these to `readonly` and initialize to `undefined` here?  I was initially confused and wondering how the promise actually got resolved since the constructor code is below this.